### PR TITLE
Updated Correction minitrees for MC compatibility

### PR DESCRIPTION
**Pull request by jpienaar13**
Changed corrections minitrees to look for a new tag in pax_metadata, 'mc_data', in order to track whether a pax file was produced from MC data or not. This is hopefully a resolution to #139.

If the file was generated from MC data, then it also attempts to pull the electron lifetime used from the pax_metadata

Hopefully this does its without breaking/affecting hax Corrections minitrees generated from real data, where the MC tag should be either non existent, or set to false if anyone does feel like adding it.

This, in addition with an update to MC/processing which I will commit in a bit, should make 
[XENON1T/pax#610](https://github.com/XENON1T/pax/pull/610) obsolete.

### CODE REVIEW

**Comment by jpienaar13** [comment]
Add @pdeperio suggestions, the electron lifetime is now stored in hax.ini to allow for separate values to be specified during fax stage (where value is specified in SR*_parameters.ini), see [XENON1T/processing#43](https://github.com/XENON1T/processing/pull/43)

**Comment by jpienaar13** [comment]
After some more input from @pdeperio, moved electron lifetime values into the correction_definitions dict in hax.ini (with some renaming, as looking for "file_names" no longer seemed entirely appropriate.)

Additionally, extended hax.runs.get_run_number(), to look in the pax_metadata for the MC tag, and if set to true, to use the MC run number. Corrections.py should now use this artificial MC run number when being made from MC pax files.

File: hax/hax.ini [hunk 1]
```diff
@@ -111,23 +111,34 @@ sc_api_url = 'https://xenon1t-daq.lngs.infn.it/slowcontrol/GetSCData'
 sc_api_username = 'slowcontrolwebserver'
 
 ##
+# MC Electron Lifetime 
+##
+#SR dependent electron lifetime defined here for minitree production.
+#NB!: Electron lifetime is also specified in simulation initialization file. These might not be the same value.
+mc_electron_lifetime_liquid = {"sr0": 450, "sr1": 550}
```
Comment (pdeperio) [inline hax/hax.ini]: Is this still needed?

File: hax/treemakers/corrections.py [hunk 2]
```diff
@@ -53,6 +53,14 @@ class Corrections(TreeMaker):
 
     extra_metadata = hax.config['corrections_definitions']
 
+    #Check if data was generated with MC.
+    #If it was, pull the electron lifetime values from the associated metadata
+    def check_for_mc(self):
+        mc_data = False
+        if 'MC' in hax.paxroot.get_metadata(self.run_name)['configuration']:
```
Comment (pdeperio) [inline hax/treemakers/corrections.py]: Can this function be moved to ```runs.py```? (Looks like maybe some duplicated code there.)

File: hax/treemakers/corrections.py [hunk 3]
```diff
@@ -92,7 +100,7 @@ def extract_data(self, event):
         # If there are no interactions cannot do anything
         if not len(event.interactions):
             return result
-
+
```
Comment (pdeperio) [inline hax/treemakers/corrections.py]: Watch out for flake8 whitespace

File: hax/treemakers/corrections.py [hunk 4]
```diff
@@ -53,6 +53,14 @@ class Corrections(TreeMaker):
...
+    #If it was, pull the electron lifetime values from the associated metadata
+    def check_for_mc(self):
+        mc_data = False
+        if 'MC' in hax.paxroot.get_metadata(self.run_name)['configuration']:
+            mc_data = hax.paxroot.get_metadata(self.run_name)['configuration']['MC']['mc_generated_data']
+        return mc_data
+
     # Electron Lifetime: hopefully doc was pulled in hax.init.
     # Otherwise get it here at significantly higher DB cost
     try:
```
Comment (pdeperio) [inline hax/treemakers/corrections.py]: Should this check be skipped for MC?

File: hax/treemakers/corrections.py [hunk 5]
```diff
@@ -144,7 +152,14 @@ def extract_data(self, event):
...
 
         # include electron lifetime correction
-        if self.elife_interpolation is not None:
+        pax_mc_data = self.check_for_mc()
+        if pax_mc_data:
+            wanted_electron_lifetime = self.get_correction("mc_electron_lifetime_liquid")
+            result['s2_lifetime_correction'] = np.exp((interaction.drift_time/1e3) /
+                                                      wanted_electron_lifetime)
+            print("This run is tagged as being MC data. Using MC electron lifetime value of %i us."
+                %wanted_electron_lifetime)
```
Comment (pdeperio) [inline hax/treemakers/corrections.py]: Add newline (flake8 issue) after here

File: hax/hax.ini (hunk 1 above)
Comment (jpienaar13) [reply hax/hax.ini]: Woops. No. Forgot to take that out.

File: hax/treemakers/corrections.py (hunk 4 above)
Comment (jpienaar13) [reply hax/treemakers/corrections.py]: I didn't want it to throw an error if it checks for MC data in the configuration settings and doesn't find it there

**Comment by jpienaar13** [comment]
I did check that I could produce corrections minitrees on real data yes.

File: hax/treemakers/corrections.py (hunk 2 above)
Comment (jpienaar13) [reply hax/treemakers/corrections.py]: Maybe. The function in runs.py, checks if it is MC data as well, but is only used to set a run number, whereas this function is used to make the decision between using the interpolated lifetime or the correction init file electron lifetime

**Comment by ErikHogenbirk** [comment]
FYI I am using this branch for some simulated AmBe data. In my simulated data, the 'MC' was not in the metadata, I suppose this will be fixed in a newer version of pax (or at the hax stage?)? I changed my local version of the code to force `mc_data = True` and then  it works like a charm. One comment, can you change the print statement (`"This run is tagged as being MC data. Using MC electron lifetime value of %i us."`) to a `log.debug`? It's printing messages for every dataset, and there are many of them for AmBe or NG MC.

**Comment by coderdj** [comment]
This looks good. But don't you also want to be able to define custom files to use for the other corrections too? Or is this not necessary?

**Comment by jpienaar13** [comment]
@ErikHogenbirk, once [XENON1T/processing#43](https://github.com/XENON1T/processing/pull/43) is merged, all future MC production should have the MC flag.

@coderdj I didn't follow your question. Which custom files where?

**Comment by pdeperio** [comment]
@coderdj i guess you mean changing the map file names in ```hax.ini```?  then i think we can pass custom ```hax.ini```s through [hax.init("custom_hax.ini")](https://github.com/XENON1T/hax/blob/967003e54d4a3acd0e02e46b9241d8ab9dfe27d9/hax/__init__.py#L29) if necessary.

**Pull request merged**